rust/src/sysusers.rs

```diff
@@ -0,0 +1,879 @@
+                    "u {} {} {} {} {}",
+                    name,
+                    uid.format_sysusers(),
+                    optional_quoted_string(&gecos),
+                    optional_quoted_string(&homedir),
+                    optional_quoted_string(&shell),
+                )
+            }
+            SysuserEntry::Group { name, gid } => format!("g {} {}", name, gid.format_sysusers()),
+            SysuserEntry::GroupMember { uname, gname } => format!("m {} {}", uname, gname),
+        }
+    }
+}
+
+/// (Partially) parse single a single line from a sysusers.d file
+fn parse_entry(line: &str) -> Fallible<PartialSysuserEntry> {
+    let err = || format_err!("Invalid sysusers entry: \"{}\"", line);
+    let raw = line.to_string();
+    // Handle the group member special case; u/g are structurually the same.
+    if line.starts_with('m') {
+        return Ok(PartialSysuserEntry::GroupMember(raw));
+    }
+    // We only parse out the type, name, and id.
+    let elts : Vec<&str> = line.split_whitespace().collect();
+    if elts.len() < 3 {
+        return Err(err())
+    }
+    let name = elts[1].to_string();
+    let id = IdSpecification::parse(elts[2])?;
+    let rest = if elts.len() > 3 {
+        elts[3..].join(" ")
+    } else {
+        "".to_string()
+    };
+    match elts[0] {
+        "u" => {
+            Ok(PartialSysuserEntry::User(PartialEntryValue { raw, name, id, rest }))
+        },
+        "g" => {
+            Ok(PartialSysuserEntry::Group(PartialEntryValue { raw, name, id, rest }))
+        },
+        "m" => {
+            // Handled above
+            unreachable!();
+        },
+        _ => Err(err()),
+    }
+}
+
+/// Parse a sysusers.d file (as a stream)
+fn parse_sysusers_stream<I: io::BufRead>(stream: I) -> Fallible<Vec<PartialSysuserEntry>> {
+    let mut res = Vec::new();
+    for line in stream.lines() {
+        let line = line?;
+        if line.starts_with("#") || line.is_empty() {
+            continue;
+        }
+        res.push(parse_entry(&line)?);
+    }
+    Ok(res)
+}
+
+fn useradd<'a, I>(args: I) -> Fallible<Vec<SysuserEntry>>
+where
+    I: IntoIterator<Item = &'a str>,
+{
+    let app = App::new("useradd")
+        .version("0.1")
+        .about("rpm-ostree useradd wrapper")
+        .arg(Arg::with_name("system").short("r"))
+        .arg(Arg::with_name("uid").short("u").takes_value(true))
+        .arg(Arg::with_name("gid").short("g").takes_value(true))
+        .arg(Arg::with_name("no-log-init").short("l"))
+        .arg(Arg::with_name("no-create-home").short("M"))
+        .arg(Arg::with_name("no-unique").short("o"))
+        .arg(Arg::with_name("groups").short("G").takes_value(true))
+        .arg(Arg::with_name("home-dir").short("d").takes_value(true))
+        .arg(Arg::with_name("comment").short("c").takes_value(true))
+        .arg(Arg::with_name("shell").short("s").takes_value(true))
+        .arg(Arg::with_name("username").takes_value(true).required(true));
```

Comment:
Hmm, I'm worried providing `useradd` and `groupadd` APIs like this might become a maintenance burden. (Though agreed in the future, packages should be moving to sysusers anyway... but legacy :( ). Even here, e.g. there's a bunch of other flags that we're not supporting, right?

Would it work to just look through the args for the ones we're interested in, and then just pass through the rest to `useradd` (and drop calling `systemd-sysusers` manually)? In the end, the crucial bit we're interested in is whether or not the invocation is specifying a specific id or not, right?

Reply:
We need to handle our fixed overrides.  I guess one way to do that would be to inject the required `-u` argument instead?

Either way though, we need to handle the arguments.